src/simplify_algebra.cpp

```diff
@@ -708,6 +710,109 @@ struct find_rsqrt
+    }
+
+    instruction_ref
+    get_reshape_ins(program& p, instruction_ref ins, std::vector<int64_t>& lens) const
+    {
+        auto outputs          = ins->outputs();
+        instruction_ref r_ins = p.end();
+        // if the reshape instruction is already there, we can reused
+        // the existing one
+        auto it = std::find_if(
+            outputs.begin(), outputs.end(), [](auto i) { return i->name() == "reshape"; });
+        if(it != outputs.end())
+        {
+            if(lens == any_cast<op::reshape>((*it)->get_operator()).dims)
+            {
+                r_ins = *it;
+            }
+        }
+
+        if(r_ins == p.end())
+        {
+            r_ins = p.insert_instruction(std::next(ins), op::reshape{lens}, ins);
+        }
+
+        return r_ins;
+    }
+
+    instruction_ref get_trans_ins(program& p, instruction_ref ins, std::vector<int64_t>& perm) const
+    {
+        auto outputs          = ins->outputs();
+        instruction_ref t_ins = p.end();
+        // if the reshape instruction is already there, we can reused
+        // the existing one
+        auto it = std::find_if(
+            outputs.begin(), outputs.end(), [](auto i) { return i->name() == "transpose"; });
+        if(it != outputs.end())
+        {
+            if(perm == any_cast<op::transpose>((*it)->get_operator()).dims)
+            {
+                t_ins = *it;
+            }
+        }
+
+        if(t_ins == p.end())
+        {
+            t_ins = p.insert_instruction(std::next(ins), op::transpose{perm}, ins);
+        }
+
+        return t_ins;
+    }
+
+    void apply(program& p, match::matcher_result r) const
+    {
+        auto ins = r.result;
+        auto slc = r.instructions["slice"];
+        auto rsp = r.instructions["reshape"];
+
+        auto tran_op               = any_cast<op::transpose>(ins->get_operator());
+        auto perm                  = tran_op.dims;
+        std::vector<int64_t> perm0 = {0, 2, 1, 3};
+        std::vector<int64_t> perm1 = {0, 2, 3, 1};
+        if(perm != perm0 and perm != perm1)
+        {
+            return;
+        }
+
+        auto slc_op               = any_cast<op::slice>(slc->get_operator());
+        std::vector<int64_t> axes = {2};
+        if(slc_op.axes != axes)
+        {
+            return;
+        }
+
+        auto starts = slc_op.starts;
+        auto ends   = slc_op.ends;
+
+        std::transform(starts.begin(), starts.end(), starts.begin(), [](auto i) { return i / 64; });
+        std::transform(ends.begin(), ends.end(), ends.begin(), [](auto i) { return i / 64; });
+
+        auto rsp_op = any_cast<op::reshape>(rsp->get_operator());
+        auto dims   = rsp_op.dims;
+        dims[2] *= 3;
+
+        auto input = slc->inputs().front();
+        auto r_ins = get_reshape_ins(p, input, dims);
+        auto t_ins = get_trans_ins(p, r_ins, perm);
+        p.replace_instruction(ins, op::slice{{1}, starts, ends}, t_ins);
```

Comment:
Why is the axis hardcoded? It could be split across any axis.

Reply:
fixed